sys/linux/dev_bifrost.txt

```diff
@@ -0,0 +1,370 @@
+ioctl$KBASE_IOCTL_MEM_PROFILE_ADD(fd fd_bifrost, cmd const[KBASE_IOCTL_MEM_PROFILE_ADD], arg ptr[in, kbase_ioctl_mem_profile_add])
+ioctl$KBASE_IOCTL_STICKY_RESOURCE_MAP(fd fd_bifrost, cmd const[KBASE_IOCTL_STICKY_RESOURCE_MAP], arg ptr[in, kbase_ioctl_sticky_resource_map])
+ioctl$KBASE_IOCTL_STICKY_RESOURCE_UNMAP(fd fd_bifrost, cmd const[KBASE_IOCTL_STICKY_RESOURCE_UNMAP], arg ptr[in, kbase_ioctl_sticky_resource_unmap])
+ioctl$KBASE_IOCTL_MEM_FIND_GPU_START_AND_OFFSET(fd fd_bifrost, cmd const[KBASE_IOCTL_MEM_FIND_GPU_START_AND_OFFSET], arg ptr[inout, kbase_ioctl_mem_find_gpu_start_and_offset])
+ioctl$KBASE_IOCTL_MEM_EXEC_INIT(fd fd_bifrost, cmd const[KBASE_IOCTL_MEM_EXEC_INIT], arg ptr[in, kbase_ioctl_mem_exec_init])
+ioctl$KBASE_IOCTL_GET_CPU_GPU_TIMEINFO(fd fd_bifrost, cmd const[KBASE_IOCTL_GET_CPU_GPU_TIMEINFO], arg ptr[inout, kbase_ioctl_get_cpu_gpu_timeinfo])
+
+ioctl$KBASE_HWCNT_READER_GET_HWVER(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_GET_HWVER], arg const[0])
+ioctl$KBASE_HWCNT_READER_GET_BUFFER_SIZE(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_GET_BUFFER_SIZE], arg const[0])
+ioctl$KBASE_HWCNT_READER_DUMP(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_DUMP], arg const[0])
+ioctl$KBASE_HWCNT_READER_CLEAR(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_CLEAR], arg const[0])
+ioctl$KBASE_HWCNT_READER_GET_BUFFER(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_GET_BUFFER], arg ptr[out, kbase_hwcnt_reader_metadata])
+ioctl$KBASE_HWCNT_READER_PUT_BUFFER(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_PUT_BUFFER], arg ptr[in, kbase_hwcnt_reader_metadata])
+ioctl$KBASE_HWCNT_READER_SET_INTERVAL(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_SET_INTERVAL], arg const[0])
+ioctl$KBASE_HWCNT_READER_ENABLE_EVENT(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_ENABLE_EVENT], arg const[0])
+ioctl$KBASE_HWCNT_READER_DISABLE_EVENT(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_DISABLE_EVENT], arg const[0])
+ioctl$KBASE_HWCNT_READER_GET_API_VERSION(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_GET_API_VERSION], arg const[0])
+
+type base_atom_id int8
+
+base_jd_atom_v2 {
+	jc		int64
+	udata		base_jd_udata
+	extres_list	int64
+	nr_extres	int16
+	compat_core_req	int16
+	pre_dep		array[base_dependency, 2]
+	atom_number	base_atom_id
+	prio		flags[base_jd_prio, int8]
+	device_nr	int8
+	jobslot		int8
+	core_req	flags[base_jd_core_req, int32]
+	renderpass_id	int8
+	padding		array[int8, 7]
+}
+
+base_dependency {
+	atom_id		base_atom_id
+	dependency_type	flags[base_jd_dep_type, int8]
+}
+
+base_jd_udata {
+	blob	array[int64, 2]
+}
+
+kbase_ioctl_job_submit {
+	addr		ptr64[out, array[base_jd_atom_v2]]
+	nr_atoms	len[addr, int32]
+# sizeof(base_jd_atom_v2)
+	stride		const[56, int32]
+}
+
+kbase_ioctl_soft_event_update {
+	event		gpu_addr
+	new_status	int32
+	flags		const[0, int32]
+}
+
+kbase_ioctl_version_check {
+	major	int16
+	minor	int16
+}
+
+kbase_ioctl_set_flags {
+	create_flags	int32
```

Comment:
should this be BASE_CONTEXT_SYSTEM_MONITOR_SUBMIT_DISABLED or BASEP_CONTEXT_MMU_GROUP_ID_MASK? kbase_api_set_flags() seems to have a check against BASEP_CONTEXT_CREATE_KERNEL_FLAGS with an early return based on this value.

Reply:
Fixed.
Related question: Since BASEP_CONTEXT_MMU_GROUP_ID_MASK is a 4-bit bit mask inside a int32, do we need to set individual bits to tell the fuzzer that each bit can be set individually?

Reply:
Hmm yes, the mask seems to be :

`0111 1000`

I'm not sure which macros correspond to each individual bit in the above mask. If you are unable to
find it as well, perhaps the following(with a comment) will suffice.

```
create_flags = flags[basep_context_create_kernel_flags, int32]
basep_context_create_kernel_flags = BASE_CONTEXT_SYSTEM_MONITOR_SUBMIT_DISABLED, 0x8, 0x10, 0x20, 0x40
```

Reply:
What Zubin suggested is probably the simplest.
If all permutations of the 4 bits are valid, then another option is to use bitfield, e.g.:
```
	create_flags	int32:4
```
Or a range:
```
	create_flags	int32[0:31]
```

Reply:
IIUC the four bits of `BASEP_CONTEXT_MMU_GROUP_ID_MASK` are used together as a 4-bit integer as some group ID, so bitfield feels more accurate to me.
I can separate that int32 into three bit fields as:
```
	create_flags	basep_context_create_kernel_flags
	...
basep_context_create_kernel_flags {
	flags0				flags[basep_context_create_kernel_flags_0, int32:3]
	basep_context_mmu_group_id	int32:4
	flags1				const[0, int32:25]
}
basep_context_create_kernel_flags_0 = BASE_CONTEXT_SYSTEM_MONITOR_SUBMIT_DISABLED
```

Also there's similar problem with `base_mem_alloc_flags`, which contains `BASE_MEM_GROUP_ID_MASK` that is also used as a 4-bit integer for some sort of "group id". But since the more significant bits after `BASE_MEM_GROUP_ID_MASK` is still used as other flags, if I using a similar approach as above, I won't be able to use existing constants for the `flags1` part...

Changed to adding individual bits into the flags for now, any suggestions?

Reply:
Options I can think of:
1. Add group id bits to the flags explicitly:
```
foo_flags = FOO1, FOO2, 1, 2, 4, 8
```
2. Shift the consts by 4 and use a bitfield:
```
foo {
    group int32:4
    flags flags[foo_flags, int32:28]
}

foo_flags = FOO1_SHIFTED, FOO2_SHIFTED

define FOO1_SHIFTED FOO1 >> 4
define FOO2_SHIFTED FOO2 >> 4
```

Reply:
I've updated the PR with option 1. (added each individual bits to the flags explicitly)